ducktape/decorate.py

```diff
@@ -0,0 +1,229 @@
+    assert 0 <= i < num_matrix_elements
+
+    indices = [0] * len(values)
+    for j in range(len(values)):
+        length_of_j = len(values[j])
+
+        indices[j] = i % length_of_j
+        i /= length_of_j
+        if i == 0:
+            break
+
+    return indices
+
+
+def matrix(**kwargs):
+    """Function decorator used to parametrize its with a matrix of values.
+    Decorating a function or method with @matrix "transforms" it into an iterable object.
+
+    Iterating yields the original function with the specified arguments injected, but there are
+    no ordering guarantees.
+
+    Example::
+
+        @matrix(x=[1, 2], y=[-1, -2])
+        def g(x, y):
+            print "x = %s, y = %s" % (x, y)
+
+        for f in g:
+            f()
+
+        # output:
+        # x = 1, y = -1
+        # x = 1, y = -2
+        # x = 2, y = -1
+        # x = 2, y = -2
+
+        # This is equivalent to the above example:
+        @matrix(x=[1, 2])
+        @matrix(y=[-1, -2])
+        def g(x, y):
+            print "x = %s, y = %s" % (x, y)
+    """
+    vars = []
+    values = []
+    for v in kwargs:
+        vars.append(v)
+        values.append(kwargs[v])
+
+    num_params = 0
+    if len(vars) > 0:
+        num_params = reduce(lambda x, y: x * len(y), values, 1)
+
+    def parametrizer(f):
+
+        kwargs_list = []
+
+        for i in range(num_params):
+            indices = i_to_indices(i, values)
+            kwargs_list.append(
+                {vars[j]: values[j][indices[j]] for j in range(len(indices))})
+
+        if _expandable(f):
+            all_kwargs_list = []
+            for sub_kwargs in f.kwargs_list:
+                for kwargs in kwargs_list:
+                    current_kwargs = kwargs.copy()
+                    current_kwargs.update(sub_kwargs)
+                    all_kwargs_list.append(current_kwargs)
+        else:
+            all_kwargs_list = kwargs_list
+
+        wrapped = TestGenerator(f, all_kwargs_list)
+        mark_as(wrapped, MATRIX)
+        return wrapped
+    return parametrizer
+
+
+def parametrize(**kwargs):
+    """Function decorator used to parametrize its arguments.
+    Decorating a function or method with @parametrize "transforms" it into an iterable object.
+
+    Iterating yields the original function with the specified arguments injected.
+
+    Example::
+
+        @parametrize(x=1, y=2 z=-1)
+        @parametrize(x=3, y=4, z=5)
+        def g(x, y, z):
+            print "x = %s, y = %s, z = %s" % (x, y, z)
+
+        for f in g:
+            f()
+
+        # output:
+        # x = 1, y = 2, z = -1
+        # x = 3, y = 4, z = 5
+    """
+    def parametrizer(f):
+        kwargs_list = [kwargs]
+        if _expandable(f):
+            kwargs_list.extend(f.kwargs_list)
+
+        wrapped = TestGenerator(f, kwargs_list)
+        mark_as(wrapped, PARAMETRIZED)
+        return wrapped
+    return parametrizer
+
+
+def _inject(**kwargs):
```

Comment:
Pretty sure this is just `functools.partial` (which is actually a bit more general since it handles positional arguments as well, although that's probably not important here).